src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -49,15 +57,153 @@ public static int TrailingZeroCount(uint value)
                 return (int)Bmi1.TrailingZeroCount(value);
             }
 
-            // Main code has behavior 0->0, so special-case in order to match intrinsic path 0->32
+            // Main code has behavior 0->0, so special-case to match intrinsic path 0->32
             if (value == 0u)
+            {
                 return 32;
+            }
 
             // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
             return Unsafe.AddByteOffset(
-                ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn), 
+                ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn),
                 // Using deBruijn sequence, k=2, n=5 (2^5=32) : 0b_0000_0111_0111_1100_1011_0101_0011_0001u
-                ((uint)((value & -value) * 0x077CB531u)) >> 27);
+                (IntPtr)(((value & -value) * 0x077CB531u) >> 27));
```

Comment:
`(IntPtr)` fixes failing units in `corefx`

Reply:
It looks like that this is using `long` -> `IntPtr` cast on 32-bit platforms. This cast is expensive - it does overflow checks that you do not need here.

Reply:
Also, the shift works on the `long` as well now - that is also expensive on 32-bit platforms.

Reply:
Can you advise on the correct fix (w.r.t. units) - the failing [test](https://ci3.dot.net/job/dotnet_corefx/job/master/job/linux-musl-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/9457/console) has error: `error CS1503: Argument 2: cannot convert from 'uint' to 'System.IntPtr'`

Reply:
`(IntPtr)(int)(((uint)((value & -value) * 0x077CB531u)) >> 27)` ?

Reply:
Ah. I misunderstood you, I thought you were trying to get rid of the explicit cast altogether. Fixed.